### PR TITLE
Feature EE-853: Remove dependency version deletes & Add unit test

### PR DESCRIPTION
**Pull request by tomirodeghiero**

### CODE REVIEW

**Comment by tomirodeghiero** [comment]
> Failed
> 
> * 0.00% Coverage on New Code (is less than 80.00%)
> 
> # Analysis Details
> ## 12 Issues
> * Bug 0 Bugs
> * Vulnerability 0 Vulnerabilities
> * Code Smell 12 Code Smells
> 
> ## Coverage and Duplications
> * Coverage 0.00% Coverage (0.00% Estimated after merge)
> * Duplications 0.54% Duplicated Code (0.10% Estimated after merge)
> 
> **Project ID:** etendosoftware_com.etendoerp.dependencymanager_AY7CyzH9DY7wyJU_mHL7
> 
> [View in SonarQube](https://sonar.etendo.cloud/dashboard?id=etendosoftware_com.etendoerp.dependencymanager_AY7CyzH9DY7wyJU_mHL7&pullRequest=5)

The 12 code smells referred to using **var** instead of **const** or **let**, so I've marked all as resolved.

**Review by Matias-Bernal — APPROVED** [review]

File: src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    String key = "group:artifact";
+    PackageDependency newDep = new PackageDependency();
+    newDep.setVersion("2.0.0");
+    dependenciesUpdate.put(key, newDep);
+
+    ChangeVersion changeVersion = new ChangeVersion();
+
+    JSONObject result = changeVersion.buildDependencyInfo(dependenciesCurrent, dependenciesUpdate, key);
+
+    assertEquals("group", result.getString("group"));
```
Comment (RubenEtendo) [inline src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java]: Use already defined constants in `PackageUtils`

File: src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    assertEquals("artifact", result.getString("artifact"));
+    assertEquals("", result.getString("version_v1"));
```
Comment (RubenEtendo) [inline src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java]: Create constants for "version_v1" and "version_v2" literals

**Review by RubenEtendo — CHANGES_REQUESTED** [review]
There are places where literals are being used instead of already defined constants in `ChangeVersion.java` and `SelectorChangeVersion.java` too

File: src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java
Comment (etendobot) [inline src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java:1]: # GPT Review for ChangeVersionTest.java
### Review

- **Estimated effort to review [1-5]:**  
2, because the changes are focused on adding a new test class with a single test method, importing necessary classes, and asserting some results.
- **Score:**  95

### Code feedback

- **File:** 
src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java
- **Language:** 
Java
- **Suggestion:** 
Consider adding more assertions to cover edge cases and improve the test coverage. It's a good practice to test various scenarios to ensure the correctness of the method under test.
- **Label:** 
enhancement
- **Existing code:**

```java
+  @Test
+  void buildDependencyInfoNewDependency() throws Exception {
+    Map<String, PackageDependency> dependenciesCurrent = new HashMap<>();
+    Map<String, PackageDependency> dependenciesUpdate = new HashMap<>();
+    String key = PackageUtil.GROUP + ":" + PackageUtil.ARTIFACT;
+    PackageDependency newDep = new PackageDependency();
+    newDep.setVersion("2.0.0");
+    dependenciesUpdate.put(key, newDep);
+
+    ChangeVersion changeVersion = new ChangeVersion();
+
+    JSONObject result = changeVersion.buildDependencyInfo(dependenciesCurrent, dependenciesUpdate, key);
+
+    assertEquals(PackageUtil.GROUP, result.getString(PackageUtil.GROUP));
+    assertEquals(PackageUtil.ARTIFACT, result.getString(PackageUtil.ARTIFACT));
+    assertEquals("", result.getString(PackageUtil.VERSION_V1));
+    assertEquals("2.0.0", result.getString(PackageUtil.VERSION_V2));
+    assertEquals(PackageUtil.NEW_DEPENDENCY, result.getString(PackageUtil.STATUS));
+  }
```
- **Improved code:**

```java
+    assertEquals(PackageUtil.GROUP, result.getString(PackageUtil.GROUP));
+    assertEquals(PackageUtil.ARTIFACT, result.getString(PackageUtil.ARTIFACT));
+    assertEquals("", result.getString(PackageUtil.VERSION_V1));
+    assertEquals("2.0.0", result.getString(PackageUtil.VERSION_V2));
+    assertEquals(PackageUtil.NEW_DEPENDENCY, result.getString(PackageUtil.STATUS));
+    // Add more assertions here to cover additional scenarios
```

File: src/com/etendoerp/dependencymanager/process/ChangeVersion.java
Comment (etendobot) [inline src/com/etendoerp/dependencymanager/process/ChangeVersion.java:1]: # GPT Review for ChangeVersion.java
### Review

- **Estimated effort to review [1-5]:**  
3, because the changes involve adding a new method, modifying existing logic, and potentially impacting the dependency management process.
- **Score:**  85

### Code feedback

- **File:** 
src/com/etendoerp/dependencymanager/process/ChangeVersion.java
- **Language:** 
Java
- **Suggestion:** 
Consider refactoring the `buildDependencyInfo` method to improve readability and maintainability. Instead of having multiple if-else conditions, consider using a switch statement or restructuring the logic to make it more concise and easier to follow. This can enhance the code's clarity and reduce the chances of errors.
- **Label:** 
maintainability
- **Existing code:**

```java
    if (depV1 == null && depV2 != null) {
      dependencyInfo.put(PackageUtil.STATUS, PackageUtil.NEW_DEPENDENCY);
    } else if (depV1 != null && depV2 != null && !StringUtils.equals(depV1.getVersion(), depV2.getVersion())) {
      dependencyInfo.put(PackageUtil.STATUS, PackageUtil.UPDATED);
    }
```
- **Improved code:**

```java
    if (depV1 == null && depV2 != null) {
      dependencyInfo.put(PackageUtil.STATUS, PackageUtil.NEW_DEPENDENCY);
    } else if (depV1 != null && depV2 != null && !StringUtils.equals(depV1.getVersion(), depV2.getVersion())) {
      dependencyInfo.put(PackageUtil.STATUS, PackageUtil.UPDATED);
    } else if (depV1 != null) {
      dependencyInfo.put(PackageUtil.STATUS, PackageUtil.DELETED);
    }
```

File: src/com/etendoerp/dependencymanager/process/SelectorChangeVersion.java
Comment (etendobot) [inline src/com/etendoerp/dependencymanager/process/SelectorChangeVersion.java:1]: # GPT Review for SelectorChangeVersion.java
### Review

- **Estimated effort to review [1-5]:**  
2, as the changes are focused and relatively easy to understand.
- **Score:**  85

### Code feedback

- **File:** 
src/com/etendoerp/dependencymanager/process/SelectorChangeVersion.java
- **Language:** 
Java
- **Suggestion:** 
Simplify the logic in `buildDependencyInfo` method by removing unnecessary conditions and handling empty strings consistently. This will improve code readability and maintainability. [important]
- **Label:** 
maintainability
- **Existing code:**

```java
    dependencyInfo.put("version_v1", depV1 != null ? depV1.getVersion() : "null");
    dependencyInfo.put("version_v2", depV2 != null ? depV2.getVersion() : "null");

    if (depV1 != null && depV2 == null) {
      dependencyInfo.put(STATUS, DELETED);
    } else if (depV1 == null && depV2 != null) {
```
- **Improved code:**

```java
    dependencyInfo.put("version_v1", depV1 != null ? depV1.getVersion() : "");
    dependencyInfo.put("version_v2", depV2 != null ? depV2.getVersion() : "");

    if (depV1 == null && depV2 != null) {
      dependencyInfo.put(STATUS, NEW_DEPENDENCY);
    } else if (depV1 != null && depV2 != null && !StringUtils.equals(depV1.getVersion(), depV2.getVersion())) {
```

File: src/com/etendoerp/dependencymanager/util/PackageUtil.java
Comment (etendobot) [inline src/com/etendoerp/dependencymanager/util/PackageUtil.java:1]: # GPT Review for PackageUtil.java
This file has no corrections to be made. Well done!

File: web/com.etendoerp.dependencymanager/js/changeVersionDropdown.js
Comment (etendobot) [inline web/com.etendoerp.dependencymanager/js/changeVersionDropdown.js:1]: # GPT Review for changeVersionDropdown.js
### Review

- **Estimated effort to review [1-5]:**  
2, as the changes are focused on updating the categorization of dependency messages and removing the 'DELETED' category.
- **Score:**  95

### Code feedback

- **File:** 
web/com.etendoerp.dependencymanager/js/changeVersionDropdown.js
- **Language:** 
JavaScript
- **Suggestion:** 
Consider simplifying the categorization of dependency messages by removing the 'DELETED' category and updating the related logic accordingly. This can help streamline the code and make it more straightforward.
- **Label:** 
simplification, best practice
- **Existing code:**

```javascript
-    var depsMessages = { 'NEW': [], 'UPDATED': [] };
+    var depsMessages = { 'NEW': [], 'UPDATED': [], 'DELETED': [] };
```
- **Improved code:**

```javascript
-    var depsMessages = { 'NEW': [], 'UPDATED': [] };
+    var depsMessages = { 'NEW': [], 'UPDATED': [] };
```

File: src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java
Comment (etendobot) [inline src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java:1]: # GPT Review for ChangeVersionTest.java
### Review

- **Estimated effort to review [1-5]:**  
2, because the changes are focused and limited to adding a new test file with test cases.
- **Score:**  95

### Code feedback

- **File:** 
src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java
- **Language:** 
Java
- **Suggestion:** 
Consider adding more assertions to cover additional scenarios in the `buildDependencyInfoNewDependency` test method. It's important to have comprehensive test coverage.
- **Label:** 
enhancement
- **Existing code:**

```java
+  @Test
+  void buildDependencyInfoNewDependency() throws Exception {
+    Map<String, PackageDependency> dependenciesCurrent = new HashMap<>();
+    Map<String, PackageDependency> dependenciesUpdate = new HashMap<>();
+    String key = PackageUtil.GROUP + ":" + PackageUtil.ARTIFACT;
+    PackageDependency newDep = new PackageDependency();
+    newDep.setVersion("2.0.0");
+    dependenciesUpdate.put(key, newDep);
+
+    ChangeVersion changeVersion = new ChangeVersion();
+
+    JSONObject result = changeVersion.buildDependencyInfo(dependenciesCurrent, dependenciesUpdate, key);
+
+    assertEquals(PackageUtil.GROUP, result.getString(PackageUtil.GROUP));
+    assertEquals(PackageUtil.ARTIFACT, result.getString(PackageUtil.ARTIFACT));
+    assertEquals("", result.getString(PackageUtil.VERSION_V1));
+    assertEquals("2.0.0", result.getString(PackageUtil.VERSION_V2));
+    assertEquals(PackageUtil.NEW_DEPENDENCY, result.getString(PackageUtil.STATUS));
+  }
```
- **Improved code:**

```java
+  @Test
+  void buildDependencyInfoNewDependency() throws Exception {
+    Map<String, PackageDependency> dependenciesCurrent = new HashMap<>();
+    Map<String, PackageDependency> dependenciesUpdate = new HashMap<>();
+    String key = PackageUtil.GROUP + ":" + PackageUtil.ARTIFACT;
+    PackageDependency newDep = new PackageDependency();
+    newDep.setVersion("2.0.0");
+    dependenciesUpdate.put(key, newDep);
+
+    ChangeVersion changeVersion = new ChangeVersion();
+
+    JSONObject result = changeVersion.buildDependencyInfo(dependenciesCurrent, dependenciesUpdate, key);
+
+    assertEquals(PackageUtil.GROUP, result.getString(PackageUtil.GROUP));
+    assertEquals(PackageUtil.ARTIFACT, result.getString(PackageUtil.ARTIFACT));
+    assertEquals("", result.getString(PackageUtil.VERSION_V1));
+    assertEquals("2.0.0", result.getString(PackageUtil.VERSION_V2));
+    assertEquals(PackageUtil.NEW_DEPENDENCY, result.getString(PackageUtil.STATUS));
+    // Add more assertions here to cover additional scenarios
+  }
```

File: src/com/etendoerp/dependencymanager/actions/InstallDependency.java
Comment (etendobot) [inline src/com/etendoerp/dependencymanager/actions/InstallDependency.java:1]: # GPT Review for InstallDependency.java
### Review

- **Estimated effort to review [1-5]:**  
2, because the changes are focused and relatively easy to understand.
- **Score:**  85

### Code feedback

- **File:** 
src/com/etendoerp/dependencymanager/actions/InstallDependency.java
- **Language:** 
Java
- **Suggestion:** 
Consider using constants for `PackageUtil.GROUP` and `PackageUtil.ARTIFACT` to improve code readability and maintainability. This can help avoid potential typos and make the code more self-explanatory.
- **Label:** 
maintainability
- **Existing code:**

```java
        .setNamedParameter("group", group)
        .setNamedParameter("artifact", artifact)
```
- **Improved code:**

```java
        .setNamedParameter(PackageUtil.GROUP, group)
        .setNamedParameter(PackageUtil.ARTIFACT, artifact)
```

File: src/com/etendoerp/dependencymanager/process/ChangeVersion.java
Comment (etendobot) [inline src/com/etendoerp/dependencymanager/process/ChangeVersion.java:1]: # GPT Review for ChangeVersion.java
### Review

- **Estimated effort to review [1-5]:**  
3, because the changes involve adding a new method `buildDependencyInfo` to the `ChangeVersion` class, modifying the logic within the method, and adjusting the `processDependencyChanges` method. The changes also include handling dependency status and updating or creating dependencies based on the comparison results.
- **Score:**  85

### Code feedback

- **File:** 
src/com/etendoerp/dependencymanager/process/ChangeVersion.java
- **Language:** 
Java
- **Suggestion:** 
Consider refactoring the `buildDependencyInfo` method to improve readability and maintainability. Extract the logic for handling dependency status into separate methods for better code organization and clarity. Ensure consistent use of constants for keys like `PackageUtil.GROUP`, `PackageUtil.ARTIFACT`, `PackageUtil.VERSION_V1`, etc. [important]
- **Label:** 
maintainability
- **Existing code:**


```java
public JSONObject buildDependencyInfo(Map<String, PackageDependency> dependenciesCurrent, Map<String, PackageDependency> dependenciesUpdate, String key) throws JSONException {
    JSONObject dependencyInfo = new JSONObject();
    PackageDependency depV1 = dependenciesCurrent.get(key);
    PackageDependency depV2 = dependenciesUpdate.get(key);

    String[] parts = key.split(":");
    dependencyInfo.put(PackageUtil.GROUP, parts[0]);
    dependencyInfo.put(PackageUtil.ARTIFACT, parts[1]);
    dependencyInfo.put(PackageUtil.VERSION_V1, depV1 != null ? depV1.getVersion() : "");
    dependencyInfo.put(PackageUtil.VERSION_V2, depV2 != null ? depV2.getVersion() : "");

    if (depV1 == null && depV2 != null) {
        dependencyInfo.put(PackageUtil.STATUS, PackageUtil.NEW_DEPENDENCY);
    } else if (depV1 != null && depV2 != null && !StringUtils.equals(depV1.getVersion(), depV2.getVersion())) {
        dependencyInfo.put(PackageUtil.STATUS, PackageUtil.UPDATED);
    }

    return dependencyInfo;
}
```

- **Improved code:**


```java
public JSONObject buildDependencyInfo(Map<String, PackageDependency> dependenciesCurrent, Map<String, PackageDependency> dependenciesUpdate, String key) throws JSONException {
    JSONObject dependencyInfo = new JSONObject();
    PackageDependency depV1 = dependenciesCurrent.get(key);
    PackageDependency depV2 = dependenciesUpdate.get(key);

    String[] parts = key.split(":");
    dependencyInfo.put(PackageUtil.GROUP, parts[0]);
    dependencyInfo.put(PackageUtil.ARTIFACT, parts[1]);
    dependencyInfo.put(PackageUtil.VERSION_V1, depV1 != null ? depV1.getVersion() : "");
    dependencyInfo.put(PackageUtil.VERSION_V2, depV2 != null ? depV2.getVersion() : "");

    if (depV1 == null && depV2 != null) {
        dependencyInfo.put(PackageUtil.STATUS, PackageUtil.NEW_DEPENDENCY);
    } else if (depV1 != null && depV2 != null && !StringUtils.equals(depV1.getVersion(), depV2.getVersion())) {
        dependencyInfo.put(PackageUtil.STATUS, PackageUtil.UPDATED);
    }

    return dependencyInfo;
}
```

File: src/com/etendoerp/dependencymanager/process/SelectorChangeVersion.java
Comment (etendobot) [inline src/com/etendoerp/dependencymanager/process/SelectorChangeVersion.java:1]: # GPT Review for SelectorChangeVersion.java
### Review

- **Estimated effort to review [1-5]:**  
2, because the changes are straightforward and focused on updating variable names and improving clarity in method documentation.
- **Score:**  95

### Code feedback

- **File:** 
src/com/etendoerp/dependencymanager/process/SelectorChangeVersion.java
- **Language:** 
Java
- **Suggestion:** 
Consider simplifying the logic in the `buildDependencyInfo` method by removing unnecessary conditions and improving variable naming for clarity. Update the method to handle cases where `depV1` is null and `depV2` is not null more effectively. 
- **Label:** 
maintainability
- **Existing code:**

```java
    if (depV1 != null && depV2 == null) {
      dependencyInfo.put(STATUS, DELETED);
    } else if (depV1 == null && depV2 != null) {
      dependencyInfo.put(STATUS, NEW_DEPENDENCY);
    } else if (depV1 != null && !StringUtils.equals(depV1.getVersion(), depV2.getVersion())) {
      dependencyInfo.put(STATUS, UPDATED);
    } else {
      return null;
    }
```
- **Improved code:**

```java
    if (depV1 == null && depV2 != null) {
      dependencyInfo.put(STATUS, NEW_DEPENDENCY);
    } else if (depV1 != null && depV2 != null && !StringUtils.equals(depV1.getVersion(), depV2.getVersion())) {
      dependencyInfo.put(STATUS, UPDATED);
    } else {
      return null;
    }
```

File: src/com/etendoerp/dependencymanager/process/UpdatePackages.java
Comment (etendobot) [inline src/com/etendoerp/dependencymanager/process/UpdatePackages.java:1]: # GPT Review for UpdatePackages.java
### Review

- **Estimated effort to review [1-5]:**  
2, because the changes are straightforward and localized to specific methods.
- **Score:**  95

### Code feedback

- **File:** 
src/com/etendoerp/dependencymanager/process/UpdatePackages.java
- **Language:** 
Java
- **Suggestion:** 
Consider using constants for the parameter names in `setNamedParameter` to improve code readability and maintainability. This can help avoid potential typos and ensure consistency.
- **Label:** 
maintainability
- **Existing code:**

```java
.setNamedParameter("group", group)
.setNamedParameter("artifact", artifact)
.setNamedParameter("version", version)
```
- **Improved code:**

```java
.setNamedParameter(PackageUtil.GROUP, group)
.setNamedParameter(PackageUtil.ARTIFACT, artifact)
.setNamedParameter(PackageUtil.VERSION, version)
```

File: src/com/etendoerp/dependencymanager/util/PackageUtil.java
Comment (etendobot) [inline src/com/etendoerp/dependencymanager/util/PackageUtil.java:1]: # GPT Review for PackageUtil.java
This file has no corrections to be made. Well done!

File: web/com.etendoerp.dependencymanager/js/changeVersionDropdown.js
Comment (etendobot) [inline web/com.etendoerp.dependencymanager/js/changeVersionDropdown.js:1]: # GPT Review for changeVersionDropdown.js
### Review

- **Estimated effort to review [1-5]:**  
2, because the changes are focused and relatively straightforward.
- **Score:**  85

### Code feedback

- **File:** 
web/com.etendoerp.dependencymanager/js/changeVersionDropdown.js
- **Language:** 
JavaScript
- **Suggestion:** 
Consider removing the unnecessary 'DELETED' category from dependency messages processing to simplify the logic and improve readability. Since the 'DELETED' category is not being used, removing it can make the code more concise and easier to understand. [important]
- **Label:** 
maintainability
- **Existing code:**

```javascript
case '[Deleted]':
    depsMessages['DELETED'].push(message);
    break;
```
- **Improved code:**

```javascript
// Removed the 'DELETED' category as it is not being used
// case '[Deleted]':
//     depsMessages['DELETED'].push(message);
//     break;
```

**Comment by RubenEtendo** [comment]
> Failed
> 
> * 0.00% Coverage on New Code (is less than 80.00%)
> 
> # Analysis Details
> ## 1 Issue
> * Bug 0 Bugs
> * Vulnerability 0 Vulnerabilities
> * Code Smell 1 Code Smell
> 
> ## Coverage and Duplications
> * Coverage 0.00% Coverage (0.00% Estimated after merge)
> * Duplications 0.00% Duplicated Code (0.10% Estimated after merge)
> 
> **Project ID:** etendosoftware_com.etendoerp.dependencymanager_AY7CyzH9DY7wyJU_mHL7
> 
> [View in SonarQube](https://sonar.etendo.cloud/dashboard?id=etendosoftware_com.etendoerp.dependencymanager_AY7CyzH9DY7wyJU_mHL7&pullRequest=5)

This issue was just a 'use let or const' issue. I already marked it as 'won't fix'

File: src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.etendoerp.dependencymanager.data.PackageDependency;
+import com.etendoerp.dependencymanager.util.PackageUtil;
+
+class ChangeVersionTest {
+  @Test
+  void buildDependencyInfoNewDependency() throws Exception {
+    Map<String, PackageDependency> dependenciesCurrent = new HashMap<>();
+    Map<String, PackageDependency> dependenciesUpdate = new HashMap<>();
+    String key = PackageUtil.GROUP + ":" + PackageUtil.ARTIFACT;
```
Comment (RubenEtendo) [inline src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java:18]: Is it correct to join both the dependency group and artifact with ":"? Shouldn't it be "."?

Comment (tomirodeghiero) [reply src-test/src/com/etendoerp/dependencymanager/process/ChangeVersionTest.java:18]: > Is it correct to join both the dependency group and artifact with ":"? Shouldn't it be "."?

@RubenEtendo In the context of the method buildDependencyInfo provided, using a colon (":") to join both the dependency group and artifact is indeed correct. The method splits the key on ":", using the first part for the group and the second for the artifact. Here's how it's done:

```java
public JSONObject buildDependencyInfo(Map<String, PackageDependency> dependenciesCurrent, Map<String, PackageDependency> dependenciesUpdate, String key) throws JSONException {
    JSONObject dependencyInfo = new JSONObject();
    PackageDependency depV1 = dependenciesCurrent.get(key);
    PackageDependency depV2 = dependenciesUpdate.get(key);

    String[] parts = key.split(":");
    dependencyInfo.put(PackageUtil.GROUP, parts[0]);
    dependencyInfo.put(PackageUtil.ARTIFACT, parts[1]);
    dependencyInfo.put(PackageUtil.VERSION_V1, depV1 != null ? depV1.getVersion() : "");
    dependencyInfo.put(PackageUtil.VERSION_V2, depV2 != null ? depV2.getVersion() : "");

    if (depV1 == null && depV2 != null) {
      dependencyInfo.put(PackageUtil.STATUS, PackageUtil.NEW_DEPENDENCY);
    } else if (depV1 != null && depV2 != null && !StringUtils.equals(depV1.getVersion(), depV2.getVersion())) {
      dependencyInfo.put(PackageUtil.STATUS, PackageUtil.UPDATED);
    }

    return dependencyInfo;
}
```

So, the ":" is used intentionally and aligns with the method's logic for parsing dependency identifiers.

**Review by RubenEtendo — APPROVED** [review]
